### PR TITLE
Memory footprint refactor

### PR DESCRIPTION
**Pull request by chrispyles**
Refactoring of memory footprints. Precursor to implementation of #123

### CODE REVIEW

**Comment by marijanbeg** [comment]
A clever idea, @chrispyles. LGTM.

**Comment by leestott** [comment]
Chris I really like the look of this 👍

**Review by leestott — APPROVED** [review]
Checks that all child annotations are satisfied by the values in ``observed_values``, and Checks that all child annotations are satisfied by the values in the memory footprint, and that the timestamps of the satisfying values occur in non-decreasing

File: pybryt/annotations/annotation.py [hunk 1]
```diff
@@ -6,6 +6,8 @@
 from collections import OrderedDict
 from typing import Any, Dict, List, Optional, Tuple
 
+from ..execution import is_complexity_tracing_enabled, MemoryFootprint
```
Comment (tonybaloney) [inline pybryt/annotations/annotation.py:9]: absolute imports would be clearer here

```python
from pybryt.execution import is_complexity_tracing_enabled, MemoryFootprint
```

File: pybryt/annotations/collection.py [hunk 2]
```diff
@@ -6,6 +6,8 @@
 
 from .annotation import Annotation, AnnotationResult
 
+from ..execution import MemoryFootprint
```
Comment (tonybaloney) [inline pybryt/annotations/collection.py:9]: absolute imports would be clearer

File: pybryt/annotations/complexity/annotation.py [hunk 3]
```diff
@@ -5,8 +5,10 @@
 from typing import Any, List, Tuple
 
 from . import complexities as cplx
+
 from ..annotation import Annotation, AnnotationResult
-from ...execution import TimeComplexityResult
+
+from ...execution import MemoryFootprint, TimeComplexityResult
```
Comment (tonybaloney) [inline pybryt/annotations/complexity/annotation.py:11]: absolute imports would be clearer (and in this module generally)

File: pybryt/__init__.py [hunk 4]
```diff
@@ -1,5 +1,6 @@
 """Auto-assessment library for Python"""
 
+from . import utils
 from .annotations import *
```
Comment (tonybaloney) [inline pybryt/__init__.py:4]: avoid star imports unless absolutely necessary

File: pybryt/annotations/value.py [hunk 5]
```diff
@@ -284,6 +282,15 @@ def check_values_equal(value, other_value, atol = None, rtol = None, equivalence
                         if len(value) != len(other_value):
                             return False
                         res = np.array(np.isclose(v, o, atol=atol, rtol=rtol) for v, o in zip(sorted(value), sorted(other_value))).all()
+                    elif isinstance(value, dict):
+                        if not isinstance(other_value, dict):
+                            return False
+                        if all(isinstance(i, numbers.Real) for i in value.values()):
+                            res = np.array(np.isclose(k1, k2) and np.isclose(v1, v2) \
+                                for (k1, v1), (k2, v2) in zip(sorted(value.items(), key=lambda t: t[0]), \
```
Comment (tonybaloney) [inline pybryt/annotations/value.py:290]: very clever :-) but someone coming into this code in the future will have to spent a while figuring out what this does. Please either refactor explode this, or add some comments to explain what you're doing

File: pybryt/execution/memory_footprint.py [hunk 6]
```diff
@@ -0,0 +1,269 @@
+"""Memory footprint container for PyBryt"""
+
+import nbformat
+
+from typing import Any, List, Optional, Set, Tuple, Union
+
+from ..utils import filter_picklable_list, pickle_and_hash
```
Comment (tonybaloney) [inline pybryt/execution/memory_footprint.py:7]: absolute imports would be clearer

**Comment by tonybaloney** [comment]
couple of comments, the absolute imports should probably be done across the board.

related question: how does Dill handle the student running code in a different version (different pickle protocol version) of Python to the teacher?

**Pull request merged**